_posts/2017-11-09-totw-135.md

```diff
@@ -0,0 +1,83 @@
+---
+title: "Tip of the Week #135: Test the Contract, not the Implementation"
+layout: tips
+sidenav: side-nav-tips.html
+published: draft
+permalink: tips/135
+type: markdown
+---
+
+# Tip of the Week #135: Test the Contract, not the Implementation
+
+Originally posted as totw/135 on ORIGINAL DATE
+
+By James Dennett (jdennett@google.com)
+
+"If you have one true friend you have more than your share." — Thomas Fuller
+
+C++ has a somewhat elaborate access control mechanism using `public`,
+`protected`, `private`, and `friend`. Test code has its own rules of etiquette
+for using those facilities, and GoogleTest augments them with its own
+`FRIEND_TEST` macro. Use of `FRIEND_TEST` should be a last resort, not a
+preferred choice.
+
+We write tests to find bugs in the implementation of a component's contract, or
+to give us sufficient confidence that there are no such bugs. When using
+test-driven development (TDD), we also write tests to help us to design that
+contract. Tests that depend on unspecified aspects of a component are brittle,
+and prone to reporting failures even when the production code is working
+correctly.
+
+Prefer to test via the public interface of a component. More generally, tests
+should verify the contract of a component and, just as with any other client,
+should not make assumptions beyond what's guaranteed.
+
+Sometimes it's hard to get sufficient coverage when testing via a minimal
+interface. If your component is implementing a very narrow interface specified
+by a base class (e.g., one with only a `ProcessItem` virtual function) and it's
+impractical to gain sufficient confidence from a test that uses only that
+interface, consider creating a new, testable component containing the
+implementation details. Then the class containing the virtual function can be so
+simple that it needs only minimal testing. BUILD visibility can be used to
+restrict use of your implementation class if needed (and if your build system
+supports it).
+
+If a test depends on just one or two private functions, consider making those
+functions part of the public interface. This isn't so bad: you'll need it to
```

Comment:
s/it/them/ twice in this sentence to make it agree with "one or two private functions" in the previous sentence.  (In the following sentence "it" is correct.)

Reply:
Done.